src/GoTrueClient.ts

```diff
@@ -530,28 +530,45 @@ export default class GoTrueClient {
   }
 
   /**
-   * Sets the session data from refresh token and returns current session or an error if the refresh token is invalid.
-   * @param refresh_token A refresh token returned by supabase auth.
+   * Sets the session data from the current session. If the current session is expired, setSession will take care of refreshing it to obtain a new session.
+   * If the refresh token in the current session is invalid and the current session has expired, an error will be thrown.
+   * If the current session does not contain at expires_at field, setSession will use the exp claim defined in the access token.
+   * @param currentSession The current session that minimally contains an access token, refresh token and a user.
    */
-  async setSession(refresh_token: string): Promise<AuthResponse> {
+  async setSession(currentSession: Session): Promise<AuthResponse> {
```

Comment:
This should work with just an `access_token` and a `refresh_token`, no? With it being typed as `Session` typescript would complain, no?

Reply:
hmm why would typescript complain? we defined the `Session` type as an interface already and i think it feels slightly more intuitive that `setSession` takes in a session object

Reply:
The session interface is 

```js
export interface Session {
  provider_token?: string | null
  /**
   * The access token jwt. It is recommended to set the JWT_EXPIRY to a shorter expiry value.
   */
  access_token: string
  /**
   * A one-time used refresh token that never expires.
   */
  refresh_token: string
  /**
   * The number of seconds until the token expires (since it was issued). Returned when a login is confirmed.
   */
  expires_in: number
  /**
   * A timestamp of when the token will expire. Returned when a login is confirmed.
   */
  expires_at?: number
  token_type: string
  user: User
}
```

Meaning `access_token`, `refresh_token`, `expires_in`, and `user` will be required to be provided. However the code itself doesn't require them. 

So I think we should change it to 

```js
async setSession(currentSession: Pick<Session, 'access_token' | 'refresh_token'> & Partial<Session>): Promise<AuthResponse>
```

This should allow us to provide an object that has only `access_token` & `refresh_token` as well as a full session object. Wdyt?

Reply:
Hmm is there any reason why the user wouldn't want to pass in a full session? i feel like this makes things clearer on the server-side because you'll always end up with a complete session. 

In v2, the `Session` type was made tighter to make it clear that one can always expect a `refresh_token` and a `user` unlike in v1. This also makes it easy because the user can just do `getSession` and pass the session returned into `setSession` immediately. 

Also, if we change the type to `Pick<Session, 'access_token' | 'refresh_token'> & Partial<Session>`, then we need to make an extra call inside `setSession` to get the user object using the `access_token` (for the case that the `access_token` hasn't expired) before we can save the session.